operator/pkg/ciliumidentity/reconciler.go

```diff
@@ -0,0 +1,474 @@
+// 3. Creates CID - If CID only exists in the desired state cache.
+// 4. Updates CID - If CIDs in the desired state cache and watcher's store are
+// not the same.
+func (r *reconciler) reconcileCID(cidResourceKey resource.Key) error {
+	cidName := cidResourceKey.Name
+	storeCIDObj, existsInStore, err := r.cidStore.CacheStore().GetByKey(cidResourceKey.Name)
+
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return err
+	}
+
+	var storeCID *cilium_api_v2.CiliumIdentity
+	if existsInStore {
+		var ok bool
+		storeCID, ok = storeCIDObj.(*cilium_api_v2.CiliumIdentity)
+		if !ok {
+			return fmt.Errorf("wrong type (%T) of object when getting CID %q from the CID store", storeCIDObj, cidName)
+		}
+	}
+
+	cidKey, existsInDesiredState := r.desiredCIDState.LookupByID(cidName)
+	if !existsInDesiredState && !existsInStore {
+		_ = r.makeIDAvailable(cidName)
+		return nil
+	}
+
+	cidIsUsed := r.cidIsUsedInPods(cidName) || r.cidIsUsedInCEPOrCES(cidName)
+	if !existsInDesiredState {
+		if cidIsUsed {
+			return nil
+		}
+		r.cidCreateLock.Lock()
+		defer r.cidCreateLock.Unlock()
+		return r.handleCIDDeletion(cidName)
+	}
+
+	if !cidIsUsed {
+		if existsInStore {
+			r.cidCreateLock.Lock()
+			defer r.cidCreateLock.Unlock()
+			return r.handleCIDDeletion(cidName)
+		}
+
+		r.desiredCIDState.Remove(cidName)
+		return nil
+	}
+
+	if !existsInStore {
+		return r.createCID(cidName, cidKey)
+	}
+
+	storeCIDKey := key.GetCIDKeyFromLabels(storeCID.SecurityLabels, "")
+	if cidKey.Equals(storeCIDKey.LabelArray) {
+		return nil
+	}
+
+	return r.updateCID(storeCID, cidKey)
+}
+
+func (r *reconciler) createCID(cidName string, cidKey *key.GlobalIdentity) error {
+	cidLabels := cidKey.GetAsMap()
+	selectedLabels, skippedLabels := identitybackend.SanitizeK8sLabels(cidLabels)
+	r.logger.Debug("Skipped non-kubernetes labels when labelling CID. All labels will still be used in identity determination", logfields.Labels, skippedLabels)
+
+	cid := &cilium_api_v2.CiliumIdentity{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   cidName,
+			Labels: selectedLabels,
+		},
+		SecurityLabels: cidLabels,
+	}
+
+	r.logger.Info("Creating CID", "labels", cidLabels, logfields.CIDName, cidName)
+
+	_, err := r.clientset.CiliumV2().CiliumIdentities().Create(context.TODO(), cid, metav1.CreateOptions{})
+	return err
+}
+
+func (r *reconciler) updateCID(cid *cilium_api_v2.CiliumIdentity, cidKey *key.GlobalIdentity) error {
+	cidLabels := cidKey.GetAsMap()
+	selectedLabels, skippedLabels := identitybackend.SanitizeK8sLabels(cidLabels)
+	r.logger.Debug("Skipped non-kubernetes labels when labelling CID. All labels will still be used in identity determination", logfields.Labels, skippedLabels)
+
+	cid.Labels = selectedLabels
+	cid.SecurityLabels = cidLabels
+
+	r.logger.Info("Updating CID", logfields.CIDName, cid.Name)
+
+	_, err := r.clientset.CiliumV2().CiliumIdentities().Update(context.TODO(), cid, metav1.UpdateOptions{})
```

Comment:
We should use the reconciler context here, same for the other usages of the clientset.

Reply:
done